File: src/bluetooth/libqlowenergyadvertisingdata.cpp
#include <QBluetoothUuid>
#include <QByteArray>
#include <QList>
#include <QLowEnergyAdvertisingData>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qlowenergyadvertisingdata.h>
#include "libqlowenergyadvertisingdata.h"
#include "libqlowenergyadvertisingdata.hxx"

QLowEnergyAdvertisingData* QLowEnergyAdvertisingData_new() {
    return new QLowEnergyAdvertisingData();
}

QLowEnergyAdvertisingData* QLowEnergyAdvertisingData_new2(const QLowEnergyAdvertisingData* other) {
    return new QLowEnergyAdvertisingData(*other);
}

void QLowEnergyAdvertisingData_OperatorAssign(QLowEnergyAdvertisingData* self, const QLowEnergyAdvertisingData* other) {
    self->operator=(*other);
}

void QLowEnergyAdvertisingData_SetLocalName(QLowEnergyAdvertisingData* self, const libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    self->setLocalName(name_QString);
}

libqt_string QLowEnergyAdvertisingData_LocalName(const QLowEnergyAdvertisingData* self) {
    QString _ret = self->localName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

uint16_t QLowEnergyAdvertisingData_InvalidManufacturerId() {
    return static_cast<uint16_t>(QLowEnergyAdvertisingData::invalidManufacturerId());
}

void QLowEnergyAdvertisingData_SetManufacturerData(QLowEnergyAdvertisingData* self, uint16_t id, const libqt_string data) {
    QByteArray data_QByteArray(data.data, data.len);
    self->setManufacturerData(static_cast<quint16>(id), data_QByteArray);
}

uint16_t QLowEnergyAdvertisingData_ManufacturerId(const QLowEnergyAdvertisingData* self) {
    return static_cast<uint16_t>(self->manufacturerId());
}

libqt_string QLowEnergyAdvertisingData_ManufacturerData(const QLowEnergyAdvertisingData* self) {
    QByteArray _qb = self->manufacturerData();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<char*>(malloc(_str.len));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    return _str;
}

void QLowEnergyAdvertisingData_SetIncludePowerLevel(QLowEnergyAdvertisingData* self, bool doInclude) {
    self->setIncludePowerLevel(doInclude);
}

bool QLowEnergyAdvertisingData_IncludePowerLevel(const QLowEnergyAdvertisingData* self) {
    return self->includePowerLevel();
}

void QLowEnergyAdvertisingData_SetDiscoverability(QLowEnergyAdvertisingData* self, int mode) {
    self->setDiscoverability(static_cast<QLowEnergyAdvertisingData::Discoverability>(mode));
}

int QLowEnergyAdvertisingData_Discoverability(const QLowEnergyAdvertisingData* self) {
    return static_cast<int>(self->discoverability());
}

void QLowEnergyAdvertisingData_SetServices(QLowEnergyAdvertisingData* self, const libqt_list /* of QBluetoothUuid* */ services) {
    QList<QBluetoothUuid> services_QList;
    services_QList.reserve(services.len);
    QBluetoothUuid** services_arr = static_cast<QBluetoothUuid**>(services.data);
    for (size_t i = 0; i < services.len; ++i) {
        services_QList.push_back(*(services_arr[i]));
    }
    self->setServices(services_QList);
}

libqt_list /* of QBluetoothUuid* */ QLowEnergyAdvertisingData_Services(const QLowEnergyAdvertisingData* self) {
    QList<QBluetoothUuid> _ret = self->services();
    // Convert QList<> from C++ memory to manually-managed C memory
    QBluetoothUuid** _arr = static_cast<QBluetoothUuid**>(malloc(sizeof(QBluetoothUuid*) * (_ret.size())));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new QBluetoothUuid(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void QLowEnergyAdvertisingData_SetRawData(QLowEnergyAdvertisingData* self, const libqt_string data) {
    QByteArray data_QByteArray(data.data, data.len);
    self->setRawData(data_QByteArray);
}

libqt_string QLowEnergyAdvertisingData_RawData(const QLowEnergyAdvertisingData* self) {
    QByteArray _qb = self->rawData();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<char*>(malloc(_str.len));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    return _str;
}

void QLowEnergyAdvertisingData_Swap(QLowEnergyAdvertisingData* self, QLowEnergyAdvertisingData* other) {
    self->swap(*other);
}

void QLowEnergyAdvertisingData_Delete(QLowEnergyAdvertisingData* self) {
    delete self;
}

File: include/bluetooth/libqlowenergyadvertisingdata.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QBluetoothUuid = @import("libqt6").QBluetoothUuid;
const qlowenergyadvertisingdata_enums = enums;
const std = @import("std");

/// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html)
pub const QLowEnergyAdvertisingData = extern struct {
    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.QLowEnergyAdvertisingData,

    pub const _is_QLowEnergyAdvertisingData = {};

    /// ### DEPRECATED: Use `new` instead
    ///
    pub const New = new;

    /// Allocate a new QLowEnergyAdvertisingData object in C++ memory
    ///
    pub fn new() QLowEnergyAdvertisingData {
        return .{ .ptr = qtc.QLowEnergyAdvertisingData_new() };
    }

    /// ### DEPRECATED: Use `new2` instead
    ///
    pub const New2 = new2;

    /// Allocate a new QLowEnergyAdvertisingData object in C++ memory
    ///
    /// ## Parameter(s):
    ///
    /// ` other: QLowEnergyAdvertisingData `
    ///
    pub fn new2(other: anytype) QLowEnergyAdvertisingData {
        comptime _ = @TypeOf(other)._is_QLowEnergyAdvertisingData;
        return .{ .ptr = qtc.QLowEnergyAdvertisingData_new2(@ptrCast(other.ptr)) };
    }

    /// ### DEPRECATED: Use `operatorAssign` instead
    ///
    pub const OperatorAssign = operatorAssign;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#operator-eq)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` other: QLowEnergyAdvertisingData `
    ///
    pub fn operatorAssign(self: QLowEnergyAdvertisingData, other: anytype) void {
        comptime _ = @TypeOf(other)._is_QLowEnergyAdvertisingData;
        qtc.QLowEnergyAdvertisingData_OperatorAssign(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### DEPRECATED: Use `setLocalName` instead
    ///
    pub const SetLocalName = setLocalName;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#setLocalName)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` name: []const u8 `
    ///
    pub fn setLocalName(self: QLowEnergyAdvertisingData, name: []const u8) void {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        qtc.QLowEnergyAdvertisingData_SetLocalName(@ptrCast(self.ptr), name_str);
    }

    /// ### DEPRECATED: Use `localName` instead
    ///
    pub const LocalName = localName;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#localName)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn localName(self: QLowEnergyAdvertisingData, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.QLowEnergyAdvertisingData_LocalName(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("QLowEnergyAdvertisingData.localName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `invalidManufacturerId` instead
    ///
    pub const InvalidManufacturerId = invalidManufacturerId;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#invalidManufacturerId)
    ///
    pub fn invalidManufacturerId() u16 {
        return qtc.QLowEnergyAdvertisingData_InvalidManufacturerId();
    }

    /// ### DEPRECATED: Use `setManufacturerData` instead
    ///
    pub const SetManufacturerData = setManufacturerData;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#setManufacturerData)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` id: u16 `
    ///
    /// ` data: []u8 `
    ///
    pub fn setManufacturerData(self: QLowEnergyAdvertisingData, id: u16, data: []u8) void {
        const data_str = qtc.libqt_string{
            .len = data.len,
            .data = data.ptr,
        };
        qtc.QLowEnergyAdvertisingData_SetManufacturerData(@ptrCast(self.ptr), @bitCast(id), data_str);
    }

    /// ### DEPRECATED: Use `manufacturerId` instead
    ///
    pub const ManufacturerId = manufacturerId;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#manufacturerId)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    pub fn manufacturerId(self: QLowEnergyAdvertisingData) u16 {
        return qtc.QLowEnergyAdvertisingData_ManufacturerId(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `manufacturerData` instead
    ///
    pub const ManufacturerData = manufacturerData;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#manufacturerData)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn manufacturerData(self: QLowEnergyAdvertisingData, allocator: std.mem.Allocator) []u8 {
        var _bytearray: qtc.libqt_string = qtc.QLowEnergyAdvertisingData_ManufacturerData(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("QLowEnergyAdvertisingData.manufacturerData: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `setIncludePowerLevel` instead
    ///
    pub const SetIncludePowerLevel = setIncludePowerLevel;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#setIncludePowerLevel)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` doInclude: bool `
    ///
    pub fn setIncludePowerLevel(self: QLowEnergyAdvertisingData, doInclude: bool) void {
        qtc.QLowEnergyAdvertisingData_SetIncludePowerLevel(@ptrCast(self.ptr), doInclude);
    }

    /// ### DEPRECATED: Use `includePowerLevel` instead
    ///
    pub const IncludePowerLevel = includePowerLevel;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#includePowerLevel)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    pub fn includePowerLevel(self: QLowEnergyAdvertisingData) bool {
        return qtc.QLowEnergyAdvertisingData_IncludePowerLevel(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `setDiscoverability` instead
    ///
    pub const SetDiscoverability = setDiscoverability;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#setDiscoverability)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` mode: qlowenergyadvertisingdata_enums.Discoverability `
    ///
    pub fn setDiscoverability(self: QLowEnergyAdvertisingData, mode: i32) void {
        qtc.QLowEnergyAdvertisingData_SetDiscoverability(@ptrCast(self.ptr), @bitCast(mode));
    }

    /// ### DEPRECATED: Use `discoverability` instead
    ///
    pub const Discoverability = discoverability;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#discoverability)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ## Returns:
    ///
    /// ` qlowenergyadvertisingdata_enums.Discoverability `
    ///
    pub fn discoverability(self: QLowEnergyAdvertisingData) i32 {
        return qtc.QLowEnergyAdvertisingData_Discoverability(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `setServices` instead
    ///
    pub const SetServices = setServices;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#setServices)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` _services: []QBluetoothUuid `
    ///
    pub fn setServices(self: QLowEnergyAdvertisingData, _services: []QBluetoothUuid) void {
        const services_list = qtc.libqt_list{
            .len = _services.len,
            .data = @ptrCast(_services.ptr),
        };
        qtc.QLowEnergyAdvertisingData_SetServices(@ptrCast(self.ptr), services_list);
    }

    /// ### DEPRECATED: Use `services` instead
    ///
    pub const Services = services;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#services)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn services(self: QLowEnergyAdvertisingData, allocator: std.mem.Allocator) []QBluetoothUuid {
        const _arr: qtc.libqt_list = qtc.QLowEnergyAdvertisingData_Services(@ptrCast(self.ptr));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QBluetoothUuid, _arr.len) catch @panic("QLowEnergyAdvertisingData.services: Memory allocation failed");
        const _data_val: [*]QtC.QBluetoothUuid = @ptrCast(@alignCast(_arr.data));
        for (0.._arr.len) |j|
            _ret[j] = .{ .ptr = _data_val[j] };
        return _ret;
    }

    /// ### DEPRECATED: Use `setRawData` instead
    ///
    pub const SetRawData = setRawData;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#setRawData)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` data: []u8 `
    ///
    pub fn setRawData(self: QLowEnergyAdvertisingData, data: []u8) void {
        const data_str = qtc.libqt_string{
            .len = data.len,
            .data = data.ptr,
        };
        qtc.QLowEnergyAdvertisingData_SetRawData(@ptrCast(self.ptr), data_str);
    }

    /// ### DEPRECATED: Use `rawData` instead
    ///
    pub const RawData = rawData;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#rawData)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn rawData(self: QLowEnergyAdvertisingData, allocator: std.mem.Allocator) []u8 {
        var _bytearray: qtc.libqt_string = qtc.QLowEnergyAdvertisingData_RawData(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("QLowEnergyAdvertisingData.rawData: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `swap` instead
    ///
    pub const Swap = swap;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#swap)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    /// ` other: QLowEnergyAdvertisingData `
    ///
    pub fn swap(self: QLowEnergyAdvertisingData, other: anytype) void {
        comptime _ = @TypeOf(other)._is_QLowEnergyAdvertisingData;
        qtc.QLowEnergyAdvertisingData_Swap(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### DEPRECATED: Use `delete` instead
    ///
    pub const Delete = delete;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#dtor.QLowEnergyAdvertisingData)
    ///
    /// Delete this object from C++ memory
    ///
    /// ## Parameter:
    ///
    /// ` self: QLowEnergyAdvertisingData `
    ///
    pub fn delete(self: QLowEnergyAdvertisingData) void {
        qtc.QLowEnergyAdvertisingData_Delete(@ptrCast(self.ptr));
    }
};

/// ### [Upstream resources](https://doc.qt.io/qt-6/qlowenergyadvertisingdata.html#public-types)
pub const enums = struct {
    pub const Discoverability = enum(i32) {
        pub const DiscoverabilityNone: i32 = 0;
        pub const DiscoverabilityLimited: i32 = 1;
        pub const DiscoverabilityGeneral: i32 = 2;
    };
};
